### PR TITLE
feat: Simplify standard permissions

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

For API scoping, we need to add a new Permission class. As I started to do this I realised that pretty much every viewset has the same set of permissions that we _always_ want to apply (they are super important!) but currently rely on the developer simply happening to implement them correctly (likely through copy-pasting.

What I want is that:
1. It is simple and obvious to have standard authentication functionality on my ViewSet without having to remember to manually apply it all
2. Ability to extend the permissions in a sensible way without having to re-implement the core logic

## Changes

* Changes the `StructuredViewSetMixin` to include permissions as well as authentication and renames it to `TeamAndOrgViewSetMixin`
* Changes the `TeamMemberAccessPermission` to behave like the `OrganizationMemberPermissions` - only validating the team access if the endpoint is itself a team-scoped endpoint
* Ensure these are applied on every viewset, with only additional permissions now needing to be specified with `permission_classes` and similarly `authentication_classes`
* Removed all duplicate logic around the permission classes
* Removes basic auth support (we don't really support it)

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/api/routing.py [hunk 1]
```diff
@@ -28,10 +31,10 @@ def __init__(self, *args, **kwargs):
         self.trailing_slash = r"/?"
 
 
-class StructuredViewSetMixin(_GenericViewSet):
+class TeamAndOrgViewSetMixin(_GenericViewSet):
     # This flag disables nested routing handling, reverting to the old request.user.team behavior
     # Allows for a smoother transition from the old flat API structure to the newer nested one
-    legacy_team_compatibility: bool = False
+    derive_current_team_from_user: bool = False
```
Comment (Twixes) [inline posthog/api/routing.py:37]: Maybe `derive_current_team_from_user_only`, because this doesn't need to be True for deriving the current team from the user – it's always possible with the explicit special value `@current`. This is only for leaving those legacy endpoints around

File: posthog/api/routing.py [hunk 2]
```diff
@@ -43,8 +46,19 @@ class StructuredViewSetMixin(_GenericViewSet):
         JwtAuthentication,
         PersonalAPIKeyAuthentication,
         authentication.SessionAuthentication,
-        authentication.BasicAuthentication,
     ]
+    additional_authentication_classes = []
+
+    permission_classes = [IsAuthenticated, OrganizationMemberPermissions, TeamMemberAccessPermission]
```
Comment (Twixes) [inline posthog/api/routing.py:52]: `TeamMemberAccessPermission` only applies to endpoints scoped by team

Comment (Twixes) [reply posthog/api/routing.py:52]: Ah, you mentioned "only validating the team access if the endpoint is itself a team-scoped endpoint" – but then why have it at all?

File: posthog/api/routing.py (hunk 1 above)
Comment (benjackwhite) [reply posthog/api/routing.py:37]: Ah good call

File: posthog/api/routing.py [hunk 3]
```diff
@@ -43,8 +46,19 @@ class StructuredViewSetMixin(_GenericViewSet):
         JwtAuthentication,
         PersonalAPIKeyAuthentication,
         authentication.SessionAuthentication,
-        authentication.BasicAuthentication,
     ]
+    additional_authentication_classes = []
```
Comment (webjunkie) [inline posthog/api/routing.py]: Idea: Why not turn it around?

If the user in the view specifies `permission_classes`, these get added. While here we have `base_permission_classes` under the hood.

Comment (benjackwhite) [reply posthog/api/routing.py]: Yeah I actually just thought about that as well :D

Comment (benjackwhite) [reply posthog/api/routing.py]: I think its annoying magic tbh. The code is already a bit hard to follow and moving even more stuff around security to some default value that is hard to trace feels error prone.

Also the new logic I changed, wouldn't support that anways.

File: ee/api/explicit_team_member.py [hunk 4]
```diff
@@ -120,7 +121,7 @@ def get_permissions(self):
         ):
             # Special case: allow already authenticated users to leave projects
             return []
-        return super().get_permissions()
+        return [permission() for permission in self.permission_classes]
```
Comment (webjunkie) [inline ee/api/explicit_team_member.py:124]: Won't this override what TeamAndOrgViewSetMixin provides?

File: posthog/api/team.py [hunk 5]
```diff
@@ -334,17 +332,16 @@ def update(self, instance: Team, validated_data: Dict[str, Any]) -> Team:
...
-class TeamViewSet(viewsets.ModelViewSet):
+class TeamViewSet(TeamAndOrgViewSetMixin, viewsets.ModelViewSet):
     """
     Projects for the current organization.
     """
 
     serializer_class = TeamSerializer
     queryset = Team.objects.all().select_related("organization")
-    permission_classes = [permissions.IsAuthenticated, PremiumMultiProjectPermissions]
+    permission_classes = [IsAuthenticated, PremiumMultiProjectPermissions]
```
Comment (webjunkie) [inline posthog/api/team.py:342]: `IsAuthenticated` is in by default, right? I'd rather remove it here to not cause confusion like thinking "wait, did I forget it elsewhere"?

File: posthog/auth.py [hunk 6]
```diff
@@ -143,8 +143,13 @@ def authenticate(self, request: Request):
             if not user.exists():
                 raise AuthenticationFailed(detail="User doesn't exist")
             return (user.first(), None)
+
         return None
 
+    # NOTE: This appears first in the authentication chain often so we want to define an authenticate_header to ensure 401 and not 403
```
Comment (webjunkie) [inline posthog/auth.py:149]: I don't know all the details behind this, but it might just be cleaner and add no new lines to subclass from `TokenAuthentication` which provides this method.

File: posthog/auth.py [hunk 7]
```diff
@@ -143,8 +143,13 @@ def authenticate(self, request: Request):
...
+    def authenticate_header(self, request: Request):
+        return "Bearer"
+
 
 class JwtAuthentication(authentication.BaseAuthentication):
```
Comment (webjunkie) [inline posthog/auth.py:154]: Actually same here, why do we not leverage `TokenAuthentication` at least a bit?

**Review by webjunkie — APPROVED** [review]
Tried to glance over all the changes and so far looks alright. Just a couple of notes.

File: ee/api/explicit_team_member.py (hunk 4 above)
Comment (benjackwhite) [reply ee/api/explicit_team_member.py:124]: Yeah there is some special cases like this where they are doing very specific things. Typically only on the "base" models like Team or Organization

File: posthog/api/team.py (hunk 5 above)
Comment (benjackwhite) [reply posthog/api/team.py:342]: Not quite - again Team is a weird one that has its own overridden get_permissions.
I still want the gains from the base class but the permissions are different here

File: posthog/auth.py (hunk 6 above)
Comment (benjackwhite) [reply posthog/auth.py:149]: I also don't know the detail and would rather not make that change here. Happy to review a follow up PR though!

Comment (benjackwhite) [reply posthog/auth.py:149]: Also - we would override almost the entire base class, only inherting a couple of properties. Doesn't feel worth it imo

File: posthog/api/team.py (hunk 5 above)
Comment (webjunkie) [reply posthog/api/team.py:342]: I see, then it makes sense.

File: posthog/api/routing.py (hunk 2 above)
Comment (Twixes) [reply posthog/api/routing.py]: So what I meant is "why have it on non-`is_team_view`" viewsets. But both work the same, so either way works

**Review by Twixes — APPROVED** [review]
LGTM!
